### DISCUSSION TITLE
SQL.Select("TOP 3")

### DISCUSSION THREAD

**Discussion by hrvoje-grabusic** [maxtoroq/DbExtensions#19] [Q&A] [answered]
Hi,
a very stupid question but i don't know how to do a select top.

I tried this
query.SELECT("TOP 3"); 
query.SELECT("i.*");
query.FROM("Projekt"] AS i");
query.WHERE("Published=''1");

and i get this string
SELECT TOP 3, i.\* FROM [Projekt] AS i WHERE Published='1'

How do i prevent the coma after top 3 ?

**Comment by maxtoroq** [comment]
``` csharp
query.SELECT("TOP 3 i.*");
```

:-)

**Comment by hrvoje-grabusic** [comment]
yes but i want to have top 3 in a if condition
if(top>0){
query.SELECT("TOP "+top); 
}
query.SELECT("i.*")

**Comment by maxtoroq** [comment, marked as answer]
You can do this:

``` csharp
if (top > 0) {
   query.SELECT("TOP {0} i.*", top); 
} else {
   query.SELECT("i.*")
}
```

There are other ways to go if you don't like the duplication of `i.*`, but you'll end up with more complicated code.

If you are using DbExtensions for mapping the you can let SqlSet handle the pagination:

``` csharp
var db = new Database("name=MyDatabase");

var set = db.From<MyClass>(query);

if (top > 0) {
   set = set.Take(top);
}

return set.AsEnumerable();
```

**Comment by hrvoje-grabusic** [comment]
Ok i just wanted to be sure i wasn't missing something.
That is still much better then writing raw sql strings.
Really awesome lib you have there the only thing i am missing with it is a more linq way to do paging.

I am using it only to build the SQL string for Kooboo CMS then i pass the sql string to the ExecuteQuery function of the CMS and It gives me back IEnumerable<Dictionary<String,Object>>

My Class extends the cms Class Textcontent which in turn extends a Dictionary so i am not sure if
it will map to my class since i keep all the values in the dictionary and not as properties. (I guess map is not the right word here)

If it does i guess i can skip the cms executeQuery function all together and make the query directly.

**Comment by maxtoroq** [comment]
You can also use SqlSet without executing the query, do this:

``` csharp
var set = db.From<MyClass>(query);

if (top > 0) {
   set = set.Take(top);
}

SqlBuilder finalQuery = set.GetDefiningQuery();
```

Then you can pass finalQuery to your CMS executeQuery function.
